hal/ia32/string.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * HAL basic routines
+ *
+ * Copyright 2012, 2021 Phoenix Systems
+ * Copyright 2001, 2005-2006 Pawel Pisarczyk
+ * Author: Pawel Pisarczyk, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "string.h"
+
+
+void *hal_memcpy(void *dst, const void *src, size_t n)
+{
+	__asm__ volatile(
+		"cld; "
+		"movl %0, %%ecx; "
+		"movl %%ecx, %%edx; "
+		"andl $3, %%edx; "
+		"shrl $2, %%ecx; "
+		"movl %1, %%edi; "
+		"movl %2, %%esi; "
+		"rep; movsl; "
+		"movl %%edx, %%ecx; "
+		"rep; movsb; "
+	:
+	: "g" (n), "g" (dst), "g" (src)
+	: "ecx", "edx", "esi", "edi", "memory", "cc");
```

Comment:
This can be simplified to:
```suggestion
	__asm__ volatile(
		"cld; "
		"rep; movsl; "
		"movl %%edx, %%ecx; "
		"rep; movsb; "
	:
	: "c" (n / 4), "d" (n & 3), "D" (dst), "S" (src)
	: "memory", "cc");
```

Reply:
I agree with proposed change - it's added in next commit.